compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1044,12 +1049,21 @@ object Types {
   	 *
      *  is approximated by constraining `A` to be =:= to `Int` and returning `ArrayBuffer[Int]`
      *  instead of `ArrayBuffer[_ >: Int | A <: Int & A]`
+     *
+     *  Exception (if `-YexplicitNulls` is set): if this type is a nullable union (i.e. of the form `T | Null`),
+     *  then the top-level union isn't widened. This is needed so that type inference can infer nullable types.
      */
     def widenUnion(implicit ctx: Context): Type = widen match {
-      case OrType(tp1, tp2) =>
-        ctx.typeComparer.lub(tp1.widenUnion, tp2.widenUnion, canConstrain = true) match {
-          case union: OrType => union.join
-          case res => res
+      case tp @ OrType(tp1, tp2) =>
+        if (ctx.explicitNulls && tp.isNullableUnion) {
+          // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
+          val OrType(leftTpe, nullTpe) = tp.normNullableUnion
```

Comment:
`widenUnion` needs to be fast, and `isNullableUnion` and `normNullableUnion` are slow. So we have to fix that. Can we formulate this method without reference to `normNullableUnion`?

Reply:
Done

Reply:
We no longer call normNullableUnion here.